File: java/complete/myretro/src/main/java/com/apress/myretro/retro/RetroBoardRepository.java
package com.apress.myretro.retro;

import com.apress.myretro.persistence.Repository;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.util.*;

@Component
public class RetroBoardRepository implements Repository<RetroBoard, UUID> {

    private Map<UUID,RetroBoard> boards = new HashMap<>() {{
        put(UUID.fromString("b193280f-7083-48f9-89b0-6826d04cb4de"),
                new RetroBoard(UUID.fromString("b193280f-7083-48f9-89b0-6826d04cb4de"), "Uno", new HashMap<>(), new HashMap<>(), LocalDateTime.now()));
        put(UUID.fromString("f73c6f55-95d9-438c-9df4-e17b123943d8"),
                new RetroBoard(UUID.fromString("f73c6f55-95d9-438c-9df4-e17b123943d8")," Dos", new HashMap<>(), new HashMap<>(),LocalDateTime.now()));
    }};

    @Override
    public RetroBoard save(RetroBoard domain) {

        if (null == domain.getId()) {
            domain.setId(UUID.randomUUID());
            this.boards.put(domain.getId(), domain);
            return domain;
        }

        RetroBoard updatedBoard = this.boards.get(domain.getId());
        updatedBoard.setName(domain.getName());
        updatedBoard.setCards(domain.getCards());
        updatedBoard.setAssignedUsers(domain.getAssignedUsers());
        updatedBoard.setModifiedAt(LocalDateTime.now());
        this.boards.put(updatedBoard.getId(),updatedBoard);
        return updatedBoard;
    }

    @Override
    public Optional<RetroBoard> findById(UUID uuid) {
        return Optional.ofNullable(this.boards.get(uuid));
    }

    @Override
    public Iterable<RetroBoard> findAll() {
        return this.boards.values();
    }

    @Override
    public void delete(UUID uuid) {
        this.boards.remove(uuid);
    }
}

File: kotlin/complete/myretro/src/main/kotlin/com/apress/myretro/retro/RetroBoardRepository.kt
package com.apress.myretro.retro

import com.apress.myretro.persistence.Repository
import com.apress.myretro.user.User
import com.apress.myretro.user.UserRole
import org.springframework.stereotype.Component
import java.time.LocalDateTime
import java.util.*
import kotlin.collections.HashMap

@Component
class RetroBoardRepository : Repository <RetroBoard,UUID> {
    var boards: MutableMap<UUID?,RetroBoard> =  mutableMapOf<UUID?,RetroBoard>(
        UUID.fromString("b193280f-7083-48f9-89b0-6826d04cb4de") to
                RetroBoard(
                    UUID.fromString("b193280f-7083-48f9-89b0-6826d04cb4de"),
                    "Uno",
                    hashMapOf( "emilio@email.com" to User("Emilio","emilio@email.com","https://www.gravatar.com/avatar/23bb62a7d0ca63c9a804908e57bf6bd4?d=wavatar","aw2s0me", listOf(
                        UserRole.USER),true)),
                    hashMapOf(),
                    LocalDateTime.now()
                ),
        UUID.fromString("f73c6f55-95d9-438c-9df4-e17b123943d8") to
                RetroBoard(
                    UUID.fromString("f73c6f55-95d9-438c-9df4-e17b123943d8"),
                    "Dos",
                    hashMapOf(),
                    hashMapOf(),
                    LocalDateTime.now()
                )
    )

    override fun save(domain: RetroBoard): RetroBoard {
        if (null == domain.id) {
            val uuid = UUID.randomUUID()
            domain.id = uuid
            boards[uuid] = domain
            return domain
        }

        val updatedBoard = boards[domain.id]
        updatedBoard?.name = domain.name
        updatedBoard?.cards = domain.cards
        updatedBoard?.assignedUsers = domain.assignedUsers
        updatedBoard?.modifiedAt = LocalDateTime.now()
        boards[updatedBoard?.id] = updatedBoard!!
        return updatedBoard!!
    }

    override fun findById(id: UUID): RetroBoard? = this.boards[id]

    override fun findAll(): Iterable<RetroBoard> = this.boards.values

    override fun delete(id: UUID) {
        this.boards.remove(id)
    }
}
